### PR TITLE
expression: constraint propagate for '>' and monotonous function

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

Propagate some thing like:

```
TO_DAYS(col) < TO_DAYS('2001-01-01') && col > '2009-04-03'  => false
```

### What is changed and how it works?

Add a propagate rule `ruleColumnGTConst` leverage this PR https://github.com/pingcap/tidb/pull/7643

Some simple rules:

```
	col > c1, col > c2, c1 > c2 => col > c1
        col > c1, col < c2, c1 >= c2 => false
```

A more complicate rule:

```
	col > c1, f(col) < c2, f is monotonous, f(c1) <= c2 => false
```

Prove:

```
        col > c1, f is monotonous => f(col) > f(c1)
        f(col) > f(c1), f(col) < c2, f(c1) >= c2 => false
```

### Check List 

Tests 

 - Unit test

Related changes

Ref https://github.com/pingcap/tidb/issues/7516

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8640)

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @eurekaka @zz-jason @winoros

File: expression/constraint_propagation.go [hunk 1]
```diff
@@ -164,3 +172,132 @@ func ruleColumnEQConst(ctx sessionctx.Context, i, j int, exprs *exprSet) {
...
+		}
+		if !isNull && v > 0 {
+			exprs.SetConstFalse()
+		}
+		return
+	}
+	return
+}
+
+var monotoneFuncs = map[string]struct{}{
```
Comment (winoros) [inline expression/constraint_propagation.go]: I think you need to point out that it's the set of monotonically increasing functions. Decresing is excluded.

File: expression/constraint_propagation.go [hunk 2]
```diff
@@ -164,3 +172,132 @@ func ruleColumnEQConst(ctx sessionctx.Context, i, j int, exprs *exprSet) {
...
+		return
+	}
+
+	// The simple case:
+	// col > c1, col < c2, c1 >= c2 => false
+	//
+	// The extended case:
+	// col > c1, f(col) < c2, f is monotonous, f(c1) <= c2 => false
+	//
+	// Prove:
```
Comment (winoros) [inline expression/constraint_propagation.go]: Proof

File: expression/constraint_propagation.go [hunk 3]
```diff
@@ -164,3 +172,132 @@ func ruleColumnEQConst(ctx sessionctx.Context, i, j int, exprs *exprSet) {
...
+	return
+}
+
+var monotoneFuncs = map[string]struct{}{
+	ast.ToDays: {},
+}
+
+// compareConstant compares two expressions. c1 and c2 should be constant with the same type.
+func compareConstant(ctx sessionctx.Context, fn string, c1, c2 Expression) (int64, bool, error) {
+	cmp, err := NewFunction(ctx, fn, c1.GetType(), c1, c2)
```
Comment (winoros) [inline expression/constraint_propagation.go]: The third param of `NewFunction` is the return type of the new function. For binary comparison function. It should be `NewFieldType(mysql.TypeTiny)`.

File: expression/constraint_propagation.go [hunk 4]
```diff
@@ -164,3 +172,132 @@ func ruleColumnEQConst(ctx sessionctx.Context, i, j int, exprs *exprSet) {
...
+	if err != nil {
+		return 0, false, errors.Trace(err)
```
Comment (winoros) [inline expression/constraint_propagation.go]: You don't need to trace it if it has been traced.

File: expression/constraint_propagation.go [hunk 5]
```diff
@@ -164,3 +172,132 @@ func ruleColumnEQConst(ctx sessionctx.Context, i, j int, exprs *exprSet) {
...
+	}
+	return
+}
+
+var monotoneFuncs = map[string]struct{}{
+	ast.ToDays: {},
+}
+
+// compareConstant compares two expressions. c1 and c2 should be constant with the same type.
+func compareConstant(ctx sessionctx.Context, fn string, c1, c2 Expression) (int64, bool, error) {
```
Comment (winoros) [inline expression/constraint_propagation.go]: What about `newFunctionAndEval(ctx, funcName, retType, args ...Expression)`?

Comment (tiancaiamao) [reply expression/constraint_propagation.go]: `newFunctionAndEval` is too general for the use here https://github.com/pingcap/tidb/pull/8640/files#diff-2244f3c3daedfac407882c798840886cR224

`compareConstant` performs a quite specific work.
@winoros

**Comment by tiancaiamao** [comment]
PTAL @winoros

**Comment by tiancaiamao** [comment]
PTAL @winoros

**Comment by tiancaiamao** [comment]
Tiny update. I remove this rule, it's not useful for partition pruning and adds code complexity

```
col > c1, col > c2, c1 > c2 => col > c1
```

PTAL @winoros

File: expression/constraint_propagation.go [hunk 6]
```diff
@@ -164,3 +171,134 @@ func ruleColumnEQConst(ctx sessionctx.Context, i, j int, exprs *exprSet) {
...
+	var col1 *Column
+	var con1 *Constant
+	col1, ok = f1.GetArgs()[0].(*Column)
+	if !ok {
+		return
+	}
+	con1, ok = f1.GetArgs()[1].(*Constant)
+	if !ok {
+		return
+	}
```
Comment (eurekaka) [inline expression/constraint_propagation.go]: We don't consider `const op col`?

Comment (tiancaiamao) [reply expression/constraint_propagation.go]: There are only two rules that really useful to partition pruning:

1. col < const
2. col >= const

Why?

```
create table ... partition by col 
p1 values less than const1
p2 values less than const2
...
```

Let's take a look at `p2`, the filter expressions are:

```
col >= const1 and col < const2
```

The final conditions would be:

```
col >= const1 and col < const2 and query
```

So the problem become that whether we can propagate:

```
col >= const1 and query => false
col < const2 and query => false
```

`ruleColumnOPConst` means use `col op const` as rule, perfectly cover the above two.

----------

You'll find the PR can't handle the `const op col`.
It doesn't matter, we can add another rule that transform
`const op col` to `col op const`

Then you'll find the rule order matters.

Not a serious problem, we can control the order by 
`newConstraintSolver(ruleTransformConstOPColumn, ruleColumnOPConst)`

I'll make pruning for  `select ... where to_days(c) > '2018-01-04'` work, and improve some cases later.

File: expression/constant_test.go [hunk 7]
```diff
@@ -32,12 +32,16 @@ var _ = Suite(&testExpressionSuite{})
 type testExpressionSuite struct{}
 
 func newColumn(id int) *Column {
```
[the remark this answers is not in the document]
Comment (tiancaiamao) [reply expression/constant_test.go]: There are so many `newColumn` in this file https://github.com/pingcap/tidb/blob/3cf386b5ca2300a0e0cebeae64d58eac9fcbff6c/expression/constant_test.go
If the function signature changes, it bring about ~50 (lines) irrelevant changes to this PR.

And this is `_test.go`, in this file `newColumn` indicates `newInt64Column`

**Pull request merged**